### PR TITLE
Add multiple selection box for results database.

### PR DESCRIPTION
**Pull request by facetothefate**
Reviewed by NOBODY (OOPS!).

* Scripts/libraries/resultsdbpy/resultsdbpy/view/static/library/js/components/TimelineComponents.js:
(XScrollableCanvasProvider):
(Timeline.CanvasSeriesComponent):

### CODE REVIEW

**Comment by facetothefate** [comment]
See the demo here
https://user-images.githubusercontent.com/4744394/140179905-f38260c1-bc64-42b0-b054-00ae501334f1.mp4

File: Tools/ChangeLog [hunk 1]
```diff
@@ -1,3 +1,13 @@
+2021-11-03  Zhifei Fang  <zhifei_fang@apple.com>
+
+        Add multiple selection box for Timeline library.
+
```
Comment (JonWBedard) [inline Tools/ChangeLog:4]: We do need a bugzilla here by policy

File: Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/library/js/components/TimelineComponents.js [hunk 2]
```diff
@@ -138,11 +149,63 @@ function XScrollableCanvasProvider(exporter, ...childrenFunctions) {
...
+    });
+    let drag = false;
+    let initPos = null;
+    const selectionBoxChangeEventStream = new EventStream();
+    selectionBoxChangeEventStream.id = 1;
+    const containerMouseDownEventStream = containerRef.fromEvent('mousedown');
+    containerMouseDownEventStream.action(e => {
+        drag = true;
+        initPos = getMousePosInCanvas(e, containerRef.element);
+        //selectionBoxRef.setState({x: initPos.x, y: initPos.y, width: 0, height: 0, show: true});
```
Comment (JonWBedard) [inline Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/library/js/components/TimelineComponents.js]: Why is this commented out?

**Review by JonWBedard — COMMENTED** [review]
I would like see this deployed to our staging instance. The video demo is super compelling, but I've found edge cases before by playing with the UI

File: Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/css/tooltip.css [hunk 3]
```diff
@@ -23,31 +23,34 @@
  * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
  
-.tooltip {
+ .tooltip {
     z-index: var(--topZIndex);
     position: absolute;
-    opacity: 80%;
```
Comment (JonWBedard) [inline Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/css/tooltip.css:29]: What is the thinking of dropping the opacity here?

File: Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/css/tooltip.css [hunk 4]
```diff
@@ -23,31 +23,34 @@
...
+    width: var(--largeSize);
+    height: var(--largeSize);
+    transform-origin: center center;
+    -webkit-backdrop-filter: blur(10px) brightness(88%);
+    backdrop-filter: blur(10px) brightness(88%);
+    color: var(--inverseColor);
 }
 
 .tooltip.arrow-up {
-     border-color: transparent transparent #cccd transparent;
```
Comment (JonWBedard) [inline Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/css/tooltip.css:39]: Why are we dropping the border-color?

File: Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js [hunk 5]
```diff
@@ -393,6 +393,8 @@ class TimelineFromEndpoint {
         });
     }
     update() {
+        if (this.radarButtonRef)
```
Comment (JonWBedard) [inline Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js:396]: This should be `issueButtonRef` or something that isn't specific to radar.

File: Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js [hunk 6]
```diff
@@ -512,6 +555,111 @@ class TimelineFromEndpoint {
...
+                    component = 'WebKit Tools';
+                break;
+                case 'internal-media-tests':
+                    component = 'WebKit Media';
+                break;//
+                case 'javascriptcore-tests':
+                    component = 'JavaScriptCore';
+                break;
+            }
+            const radarUrlComponents = {
```
Comment (JonWBedard) [inline Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js:650]: I think we need to create a more generic API for this.

OpenSource clearly needs to generate bugzilla links instead of radars. Also, things like component setting belong in the application layer, not the library, which will complicate things.

First step for generalizing this is probably to create a generic "createBug" function template taking `title`, `description`, and `classification` for sure, maybe `component` too? You will also probably need to provide some kind of filtering callback so the application instantiating the library can specify it's own components.

File: Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/tooltip.js [hunk 7]
```diff
@@ -79,11 +80,11 @@ class _ToolTip {
                         const rightPoint = stateDiff.points.length > 1 && stateDiff.points[1].x > stateDiff.points[0].x ? stateDiff.points[1] : stateDiff.points[0];
 
                         direction = 'left';
-                        let tipX = leftPoint.x - 12 - bounds.width;
```
Comment (JonWBedard) [inline Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/tooltip.js:82]: This change seems unrelated to the rest of the PR.

**Review by JonWBedard — CHANGES_REQUESTED** [review]
Seems like `TimelineComponents.js` is ready to go, the bits about actually generating the radar need some work though, because we need to support bugzilla as well.

File: Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/css/tooltip.css (hunk 3 above)
Comment (facetothefate) [reply Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/css/tooltip.css:29]: This is for changing the tooltip arrow, I made it use transform instead of border. so it will be a real arrow not a invisible square. The invisible part of that square will cover the dot, so the user cannot click on it.

File: Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/css/tooltip.css (hunk 4 above)
Comment (facetothefate) [reply Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/css/tooltip.css:39]: ditto.

File: Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js (hunk 6 above)
Comment (facetothefate) [reply Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js:650]: This is application layer already. this is not part of library

File: Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js (hunk 5 above)
Comment (facetothefate) [reply Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js:396]: I think we will need another bugzilla button ref as well, so I make this radar button.

In my design, we will have two different button, one is for radar and one is for bugzilla

File: Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/tooltip.js (hunk 7 above)
Comment (facetothefate) [reply Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/tooltip.js:82]: This is also for the new tooltip arrow

File: Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js (hunk 5 above)
Comment (JonWBedard) [reply Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js:396]: Then I think we need a list of button refs.

Presumably, we'll be adding GitHub Issues at some point in the future.

File: Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js (hunk 6 above)
Comment (JonWBedard) [reply Tools/Scripts/libraries/resultsdbpy/resultsdbpy/view/static/js/timeline.js:650]: I guess we actually have three layers in this application:
- JavaScript libraries
- resultsdbpy "library" that is a packaged application configured by the caller
- Cloud hosted specific application
I'm saying that this code setting components based on test suites belong a level up in the "cloud-hosted specific application." Otherwise this would be the first part of resultsdbpy that references a specific suite.

**Review by JonWBedard — CHANGES_REQUESTED** [review]

**Comment by facetothefate** [comment]
see #177 instead, it will be much easier to see the new change
